smalltalksrc/VMMakerTests/VMSpurEphemeronsAlgorithmTest.class.st

```diff
@@ -0,0 +1,361 @@
+	memory fullGC.
+
+	"ephemeronObjectOopOne := memory remapObj: ephemeronObjectOopOne.
+	ephemeronObjectOopTwo := memory remapObj: ephemeronObjectOopTwo."
+	
+	self assert: memory dequeueMourner equals: ephemeronObjectOopTwo.
+	self assert: memory dequeueMourner equals: ephemeronObjectOopOne
+]
+
+{ #category : #'tests - ephemerons with same key' }
+VMSpurEphemeronsAlgorithmTest >> testEphemeronsWithSameKeyNewSpace [
+
+	ephemeronObjectOopOne := self newEphemeronObject.
+	ephemeronObjectOopTwo := self newEphemeronObject.
+	nonEphemeronObjectOopOne := self newZeroSizedObject.
+
+	"Both ephemerons share the same key"
+	memory storePointer: 0
+		ofObject: ephemeronObjectOopOne
+		withValue: nonEphemeronObjectOopOne.
+	memory storePointer: 0
+		ofObject: ephemeronObjectOopTwo
+		withValue: nonEphemeronObjectOopOne.
+
+	self keepObjectInVMVariable1: ephemeronObjectOopOne.
+	self keepObjectInVMVariable2: ephemeronObjectOopTwo.
+
+	"Collect ephemerons"
+	memory doScavenge: 1.
+
+	self assert: ({ ephemeronObjectOopOne. ephemeronObjectOopTwo } includes: memory dequeueMourner).
+	self assert: ({ ephemeronObjectOopOne. ephemeronObjectOopTwo } includes: memory dequeueMourner)
+]
+
+{ #category : #'tests - ephemerons with same key' }
+VMSpurEphemeronsAlgorithmTest >> testEphemeronsWithSameKeyNewSpaceFlush [
+
+	ephemeronObjectOopOne := self newEphemeronObject.
+	ephemeronObjectOopTwo := self newEphemeronObject.
+	nonEphemeronObjectOopOne := self newZeroSizedObject.
+
+	"Both ephemerons share the same key"
+	memory storePointer: 0
+		ofObject: ephemeronObjectOopOne
+		withValue: nonEphemeronObjectOopOne.
+	memory storePointer: 0
+		ofObject: ephemeronObjectOopTwo
+		withValue: nonEphemeronObjectOopOne.
+
+	self keepObjectInVMVariable1: ephemeronObjectOopOne.
+	self keepObjectInVMVariable2: ephemeronObjectOopTwo.
+
+	"Collect both non ephemeron objects"
+	memory flushNewSpace.
+
+	self assert: ({ ephemeronObjectOopOne. ephemeronObjectOopTwo } includes: memory dequeueMourner).
+	self assert: ({ ephemeronObjectOopOne. ephemeronObjectOopTwo } includes: memory dequeueMourner)
+]
+
+{ #category : #'tests - ephemerons with same key' }
+VMSpurEphemeronsAlgorithmTest >> testEphemeronsWithSameKeyNewSpaceOldSpace [
+
+	ephemeronObjectOopOne := self newEphemeronObject.
+	ephemeronObjectOopTwo := self newEphemeronObject.
+	nonEphemeronObjectOopOne := self newOldSpaceObjectWithSlots: 0.
+
+	"Both ephemerons share the same key"
+	memory storePointer: 0
+		ofObject: ephemeronObjectOopOne
+		withValue: nonEphemeronObjectOopOne.
+	memory storePointer: 0
+		ofObject: ephemeronObjectOopTwo
+		withValue: nonEphemeronObjectOopOne.
+
+	self keepObjectInVMVariable1: ephemeronObjectOopOne.
+	self keepObjectInVMVariable2: ephemeronObjectOopTwo.
+
+	"Collect ephemerons"
+	memory fullGC.
+
+	self assert: ({ ephemeronObjectOopOne. ephemeronObjectOopTwo } includes: memory dequeueMourner).
+	self assert: ({ ephemeronObjectOopOne. ephemeronObjectOopTwo } includes: memory dequeueMourner)
+]
```

Comment:
This test is indeed missing in our suite, but this should be added to `VMSpurOldSpaceGarbageCollectorTest`?